pkg/logtailing/tailer.go

```diff
@@ -0,0 +1,212 @@
+package logtailing
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/logrusorgru/aurora"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/stripe/stripe-cli/pkg/ansi"
+	"github.com/stripe/stripe-cli/pkg/stripeauth"
+	"github.com/stripe/stripe-cli/pkg/websocket"
+)
+
+const outputFormatJSON = "json"
+
+// LogFilters contains all of the potential user-provided filters for log tailing
+type LogFilters struct {
+	FilterAccount        []string `json:"filter_account,omitempty"`
+	FilterIPAddress      []string `json:"filter_ip_address,omitempty"`
+	FilterHTTPMethod     []string `json:"filter_http_method,omitempty"`
+	FilterRequestPath    []string `json:"filter_request_path,omitempty"`
+	FilterRequestStatus  []string `json:"filter_request_status,omitempty"`
+	FilterSource         []string `json:"filter_source,omitempty"`
+	FilterStatusCode     []string `json:"filter_status_code,omitempty"`
+	FilterStatusCodeType []string `json:"filter_status_code_type,omitempty"`
+}
+
+// Config provides the configuration of a log tailer
+type Config struct {
+	APIBaseURL string
+
+	// DeviceName is the name of the device sent to Stripe to help identify the device
+	DeviceName string
+
+	// Filters for API request logs
+	Filters *LogFilters
+
+	// Key is the API key used to authenticate with Stripe
+	Key string
+
+	// Info, error, etc. logger. Unrelated to API request logs.
+	Log *log.Logger
+
+	// Force use of unencrypted ws:// protocol instead of wss://
+	NoWSS bool
+
+	// Output format for request logs
+	OutputFormat string
+
+	// WebSocketFeature is the feature specified for the websocket connection
+	WebSocketFeature string
+}
+
+// Tailer is the main interface for running the log tailing session
+type Tailer struct {
+	cfg *Config
+
+	stripeAuthClient *stripeauth.Client
+	webSocketClient  *websocket.Client
+
+	interruptCh chan os.Signal
+}
+
+// EventPayload is the mapping for fields in event payloads from request log tailing
+type EventPayload struct {
+	CreatedAt string `json:"created_at"`
```

Comment:
Are we sending `created_at` as a human-readable string? I would have expected it to be a timestamp so we could do the conversion client-side, using the local timezone.

Reply:
We do the conversion on the backend so it's sent as human-readable but in UTC, and it's noted as UTC in the output to the user.

I did this originally since I figured it's easier to just have everything that arrives on the client be in the proper form of output, so it's just `receive -> display`, as opposed to `receive -> any conversions -> display`, since doing the conversion would assumably be easier in Ruby than in Go.

If it's a strong preference to do the conversion on the client-side for the timezone to be in local time, I can make that change, but since it's also a backend change, it may delay the release of log tailing by a day.

Thoughts? @ob-stripe

Reply:
If it's not too much toil I would prefer we abide by the existing API conventions and send all times as Unix timestamps, and do any conversion client-side.

It's fairly easy to do the conversion in Go: https://play.golang.org/p/W18lIi9wDvH